src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -59,5 +63,205 @@ public void TestSignaturesV2_3_X()
+        {
+            // Temporary folder to put our test stream input.
+            using var srcTempDirectory = new TemporaryDirectory();
+            string streamInputPath = Path.Combine(srcTempDirectory.Path, "streamInput");
+
+            // [1, 2, ..., 99]
+            _spark.Range(1, 100).Write().Json(streamInputPath);
+
+            // Test a scenario where ForeachWriter runs without issues.
+            // If everything is working as expected, then:
+            // - Triggering stream will not throw an exception
+            // - 3 CSV files will be created in the temporary directory.
+            // - 0 Exception files will be created in the temporary directory.
+            // - The CSV files will contain valid data to read, where the
+            //   expected entries will contain [101, 102, ..., 199]
+            TestAndValidateForeach(
+                streamInputPath,
+                new TestForeachWriter(),
+                false,
+                3,
+                3,
+                0,
+                Enumerable.Range(101, 99));
+
+            // Test scenario where IForeachWriter.Open returns false.
+            // When IForeachWriter.Open returns false, then IForeachWriter.Process
+            // is not called.  Verify that:
+            // - Triggering stream will not throw an exception
+            // - 3 CSV files will be created in the temporary directory.
+            // - 0 Exception files will be created in the temporary directory.
+            // - The CSV files will not contain valid data to read.
+            TestAndValidateForeach(
+                streamInputPath,
+                new TestForeachWriterOpenFailure(),
+                false,
+                3,
+                3,
+                0,
+                Enumerable.Empty<int>());
+
+
+            // Test scenario where ForeachWriter.Process throws an Exception.
+            // When IForeachWriter.Process throws an Exception, then the exception
+            // is rethrown by ForeachWriterWrapper. We will limit the partitions
+            // to 1 to make validating this scenario simpler. Verify that:
+            // - Triggering stream throws an exception.
+            // - 1 CSV file will be created in the temporary directory.
+            // - 1 Exception will be created in the temporary directory. The
+            //   thrown exception from Process() will be sent to Close().
+            // - The CSV file will not contain valid data to read.
+            TestAndValidateForeach(
+                    streamInputPath,
+                    new TestForeachWriterProcessFailure(),
+                    true,
+                    1,
+                    1,
+                    1,
+                    Enumerable.Empty<int>());
+        }
+
+        private void TestAndValidateForeach(
+            string streamInputPath,
+            TestForeachWriter foreachWriter,
+            bool foreachThrows,
+            int partitions,
```

Comment:
can't you just use `expectedCSVFiles` instead? (or use `partitions` in place of `expectedCSVFiles`)

Reply:
Yeah, sounds good.